enhancements/authentication/direct-external-oidc-provider.md

```diff
@@ -0,0 +1,293 @@
+
+Currently, any component that needs to obtain tokens or authenticate users does so using the built-in OAuth server. In order to integrate and use an external OIDC Identity Provider directly, any component must replace its configuration to call the external OIDC provider instead of the built-in server. The core components that send requests to the OAuth server are:
+
+- OpenShift Console calls the OAuth server for user login to the console, and to obtain and display API access tokens
+- `oc` calls the OAuth server for user login to the API, and to obtain API access tokens
+- kube-apiserver calls the oauth-apiserver via an authentication webhook for token validation
+
+OCP provides means of dynamic OAuth2 client registration, which means that other components using the OAuth2 server might also exist; however these cases are not within the scope of this proposal.
+
+Note that the kube-apiserver already supports direct external OIDC providers; this proposal describes the mechanisms that need to be implemented in order to enable the configuration of an external OIDC provider for the kube-apiserver in an OCP cluster.
+
+To enable configuration changes for each of the core components, the Authentication CRD has been extended with a new API that allows the specification of the details of the external OIDC provider to use. For `oc` in particular, this specification must be carried out via relevant command-line options.
+
+Additionally, when an external OIDC provider is configured, any components and resources that are related to the built-in OAuth server must be removed (and recreated when the built-in OAuth server is configured anew). These components and resources are managed by the cluster-kube-apiserver-operator and the cluster-authentication-operator.
+
+### Workflow Description
+
+To use an external OIDC provider in core components, the user must update the Authentication CR and specify the provider's details in the respective fields. To use the provider with the `oc` CLI tool, the user must use the command-line flags of the tool in order to specify the provider's details.
+
+#### External OIDC provider configuration
+
+Apart from the provider URL, which is always required, the configuration details of an external OIDC provider might also include, depending on the workflow:
+
+- the ID of the corresponding OIDC client at the provider's side
+- the client secret
+- the provider's certificate authority bundle
+- any relevant extra scopes
+
+#### Authentication Resource
+
+The cluster's Authentication CR (`authentication.config/cluster`) must be modified and updated with the configuration of the external OIDC provider in the `OIDCProviders` field.
```

Comment:
```suggestion
The cluster's Authentication CR (`authentication.config/cluster`) must be modified and updated with the configuration of the external OIDC provider in the `OIDCProviders` field. The `WebhookTokenAuthenticator` field should be removed (by setting it `null`).
```
I add this explicitly because readers who are not aware of this may just set `oidcProviders` and ignore `webhookTokenAuthenticator`. This makes readers who attempt to run `oc patch` or `oc edit` on `authentication.config/cluster` the first time get:
```
The Authentication "cluster" is invalid: spec.webhookTokenAuthenticator: Invalid value: v1.WebhookTokenAuthenticator{KubeConfig:v1.SecretNameReference{Name:"webhook-authentication-integrated-oauth"}}: this field cannot be set with the "OIDC" .spec.type
```